### DISCUSSION TITLE
How to access parts of context - test suite (filename) name - from custom environment?

### DISCUSSION THREAD

**Discussion by gladykov** [vitest-dev/vitest#4979] [Q&A] [answered]
I have 2 test files:

`debug.test.ts`
`debug2.test.ts`

Can I access names/paths of those files inside custom environment `setup` and `teardown` methods? Or any other parts of Test Context? I looked on `global` object but nothing looked promising.

```
export default <Environment>{
    name: 'custom',
    transformMode: 'ssr',
    setup(global, { custom }) {
        
        global.testEnvironment = 'custom'
        global.option = custom.option
        return {
            teardown() {
                delete global.testEnvironment
                delete global.option

                if (global.__exists)
                    log('should not log')
            },
        }
    },
}
```

Use case: I record one video for whole test suite and rename video, to have name of test suite.

**Comment by sheremet-va** [comment]
You cannot, there is no connection between an environment and test files. It's possible to run several tests in the same environment.

**Reply by gladykov** [reply]
Thank you for the answer. As a workaround I tried assign context to `global` in `beforeAll`, and then access it from `teardown` . But since several test suites run in same env, only value from last executed is there. 

Is there a possibiity then to spawn separate test environment per test suite?

**Comment by hi-ogawa** [comment]
> Use case: I record one video for whole test suite and rename video, to have name of test suite.

Could you elaborate the use case further? From this description itself and your sample code, I'm not sure how custom environment is strictly relevant and it might end up an xy-problem.

For example, have you considered using custom reporter (+ maybe custom metadata) to achieve something similar? It might allow you to customize/extend certain behavior after test/suite is finished though it has some limitations.

- https://vitest.dev/advanced/metadata.html#task-metadata
- https://vitest.dev/advanced/reporters.html

**Reply by gladykov** [reply]
Sure, let me explain more:

When test suite is being run ex. `featureA.test.ts` , I spawn one browser instance per test suite. Also I start to record video, which has random name, ex. `760d74cd498488fa7bd3d67bee9b08be.webm` (thank you Playwright)

So this is Environemnt part.

Once test suite will stop running I want to:
- rename video from `760d74cd498488fa7bd3d67bee9b08be.webm` to `featureA.test.ts`
- teardown browser

This is why I thought environemnt `teardown()` is best place to do it. 

Be advised that my thought patterns are skewed by Jest ;)

**Reply by hi-ogawa** [reply, marked as answer]
Thanks for expanding on your use case. I see that's an interesting idea of using custom environment.

I don't have a full picture yet, but just an idea. If spawning playwright for each test file is a key, then have you considered using `setupFiles` + `beforeAll`?

I haven't used playwright's video feature myself, but what I'm imagining is something like this:

```ts
//
// some.setup.ts (you can access ctx.filepath for current test file path)
//
beforeAll(async (ctx) => {
  // spawn browser
  const page = await chromium.launch({ video: "on" });

  // expose to test files via global
  Object.assign(globalThis, { page });

  // teardown
  return async () => {
    const path = await page.video().path();
    renameFile(path, ctx.filepath);
    await page.close();
  }
})

//
// some.test.ts
//
test("navigate", async () => {
  await page.goto("https://vitest.dev/")
})
```

**Reply by gladykov** [reply]
Thank you @hi-ogawa this is very good solution ! I think this could work - `Environment` would provide one global browser process instance, and `setupFiles` would deal with creating new context (new private window) and renaming video at the end. Was just not aware that method returned by `beforeAll` would be executed at the end of the test suite. If we go this route, will post results

**Reply by gladykov** [reply]
Thank you, it worked! In `setupFiles.ts`

```
beforeAll(async (ctx) => {
    const testPath = ctx.file.filepath
    const testSuiteName = testPath[testPath.length - 1].split('.')[0];

    // teardown
    return async () => {
        await renameVideo(video, testSuiteName);

```
